text/0000-codify-infrastructure.md

```diff
@@ -0,0 +1,392 @@
+
+## Detailed design 🔎
+
+I wrote this out in the order that makes sense to me, but feel free to skip around. I'll go over my understanding of what comprises the "datafruits stack", a brief overview of just the critical things you need to understand Kubernetes ("just enough to be dangerous"), a look at how I think we should structure the infrastructure both inside and outside the cluster, and a gradual adoption plan to transition datafruits over to the new ecosystem.
+
+### What exactly _is_ the datafruits stack? 🤔
+
+Basically, to run on kubernetes, you need to make all of your applications (like, code that executes and runs things) [containerizable](https://opensource.com/resources/what-docker), and you need to keep your data elsewhere. We'll get into why this is the case later, but for now I'll go into these talking about the current way datafruits runs things and how we'll fit them into the new model.
+
+As near as I can tell, this is an overview of the entire tech stack that goes into making datafruits tick. I'm suuuuper shaky on this, so feedback is highly appreciated -- I'm almost certainly missing things 🙃
+
+#### Client Site (datafruits.fm) 🍉
+
+The most important (arguably) part of the whole thing is the website. The repo lives at [datafruits/datafruits](https://github.com/datafruits/datafruits). It's built on [ember.js](https://emberjs.com/), and importantly with [ember-cli-fastboot](https://github.com/ember-fastboot/ember-cli-fastboot). This means that the web server is a [super lightweight Node.js process](https://github.com/ember-fastboot/fastboot-app-server#ember-fastboot-app-server), which is trivial to containerize. More on containerization later.
+
+#### Admin Site (streampusher.com) 🚂
+
+This is the place you go where if you're Tony, you can look at all the DJs and their information, or if you're a DJ you can manage your own stuff and upload sets and things like that. I'm not _super_ sure how it works, because I'm not a DJ and I don't have access to it, and I've never been invested enough to really dig through the code.
+
+That said, it has its entire own codebase, which lives in [streampusher/frontend](https://github.com/streampusher/frontend). This, too, is an Ember app, which does NOT use ember-cli-fastboot. It currently lives on [netlify](https://www.netlify.com/), which is like a [serverless](https://aws.amazon.com/serverless/) [static site](https://en.wikipedia.org/wiki/Static_web_page) hosting platform, kind of like [GitHub Pages](https://pages.github.com/). I think. I've never used netlify, although I certainly have seen people sing its praises.
+
+At least, I think it lives on netfliy. There's a netlify config in the [`ember-cli-build.js`](https://github.com/streampusher/frontend/blob/master/ember-cli-build.js). Either way, it should be similarly easy to containerize this app the same way we'd containerize the client site.
```

Comment:
I don't think I have plans to ever convert the streampusher frontend to fastboot...there is nothing really you can see on the site unless you login, so there's no concerns for SEO or anything.

Would a static site really benefit from being containerized? It seems like it would be more work in a way, we'd have to setup some kind of server to get it running. 

Is there some other way to serve a static site that is more in line with your new infrastructure proposal? Netlify is super simple to use right now, and while I'm not opposed to using something else, all I need to do is push to master to deploy now, and its up in a minute, and I'd like to keep that kind of workflow if possible.

Reply:
yea absolutely! netlify is basically a static file server, so we could just set up CI to deploy the files to object storage and configure the DNS to point to it, digitalocean would be basically a drop-in replacement

Reply:
Right, I used to do this with S3. DO Spaces isn't exactly a CDN though, afaik, is there something more sophisticated CDN-like we could use?

Reply:
Right, it's a little more involved than just pointing it at a bucket -- [here](https://www.digitalocean.com/docs/spaces/#cdn) they talk about how to set it up. Basically, you replace <spacename>.<region>.digitaloceanspaces.com with <spacename>.<region>.*cdn*.digitaloceanspaces.com to activate it, and then you can [assign a custom subdomain](https://www.digitalocean.com/docs/spaces/how-to/customize-cdn-endpoint/) if you've registered the domain name with digital ocean.